### DISCUSSION TITLE
define the block/procedure parameters after staircase

### DISCUSSION THREAD

**Discussion by dokato** [jspsych/jsPsych#866] [Q&A] [answered]
I wonder how to implement the following scenario.
I have experiment with staircase procedure that helped me determine the parameter `my_param`.

Now I want to use that parameter in defining the block of my main experiment, but I noticed that instead of using the value that staircase gave me, it uses the initial value from the staircase procedure. Now that I look at the code it makes sense, as the variable procedure is being defined at the beginning of the browser session.

But here my intuition fails: How to update dynamically the `block` or actually just variable `conditions` that procedure iterates over after staircase?

```js
// ------ Staircase procedure -------
...

timeline.push(staircase); // this changes *my_param*

// -------- main experiment --------
var block = {
    difficulty: jsPsych.timelineVariable('diff')
   ....
};
conditions = [
{..., diff: my_param+1},
{..., diff: my_param-2},
...
]
var procedure = {
    timeline: [
        fixation,
        block,
        iti
    ],
    timeline_variables: conditions, // this should use *my_param* after staircase procedure ?!
    sample: {
        type: 'fixed-repetitions',
        size: 10,
    }
};
timeline.push(procedure);
```

**Comment by legatrix** [comment, marked as answer]
Instead of just 

```
var block = {
    difficulty: jsPsych.timelineVariable('my_param')
   ....
};
```

don't you need an array of timeline variables first, like so:

```
var conditions = [
		{difficulty: my_param}
	]

var block = {
    difficulty: jsPsych.timelineVariable('difficulty')
   ....
};
```

Otherwise, what is your `conditions` referring to?
Please ignore / delete if this is unhelpful.

**Reply by dokato** [reply]
Sorry @legatrix maybe my question wasn't clear. So yes, that's obvious that `conditions` has to have reference to `my_param`. The problem is that `my_param` value changes during the staircase and I want to use this new **updated** value in my `procedure`. I updated the question to make it clearer. Any idea how to achieve that?

**Reply by legatrix** [reply]
Sorry, my fault. Here's an idea that might help. It's a simple example which uses the data from the intro portion (in this case RT) to do something in the main portion (in this case just display it on the screen); should be self-explanatory:

```
// ------ Staircase procedure -------
// ...

timeline = [];

// timeline.push(staircase); // this changes *my_param*

var intro = {
          type: "html-keyboard-response",
          stimulus: "test",
          choices: ['f','j'],
          on_finish: function(data){
            data.my_param = data.rt;
          }
        }

timeline.push(intro);

// -------- main experiment --------


var block1 = {
          type: "html-keyboard-response",
          stimulus: function(){
            return jsPsych.data.get().select('my_param').values[0] + 500
          },
          choices: ['f','j'],
          stimulus_duration: 2000,
          post_trial_gap: 1000
        };

var block2 = {
          type: "html-keyboard-response",
          stimulus: function(){
            return jsPsych.data.get().select('my_param').values[0] - 500
          },
          choices: ['f','j'],
          stimulus_duration: 2000,
          post_trial_gap: 1000
        };

timeline.push(block1, block2);

        jsPsych.init({
            timeline: timeline,
            on_finish: function(){
                jsPsych.data.displayData();
            }
        });
```

**Reply by dokato** [reply]
🤔 so if I'm reading your example correctly in may case that would require introducing dummy trial that would take values from Staircase and assign it to `jsPsych.data`:

```js
var assigner = {
          type: "html-keyboard-response",
          stimulus: "hidden",
          trial_duration: 100;
          on_finish: function(data){
            data.my_param = staircase.my_param;
          }
        }
```

~~and then defining my conditions for randomized procedure with sth like:~~

```js
conditions = [
{..., diff: jsPsych.data.get().select('my_param').values[0] + 1},
{..., diff: jsPsych.data.get().select('my_param').values[0] - 2},
...
]
```

**Reply by dokato** [reply]
UPDATE: ahh not really
in this case `jsPsych.data.get().select('my_param').values[0]` will be undefined at the beggining of the session again.

But I think I see where you're getting at. Unfortunately, I have quite a few blocks to run, so your idea would be pretty ineffective in my case. Still looking for a solution that makes this procedure with `timeline_variables` working...

```
var procedure = {
    timeline: [
        fixation,
        block,
        iti
    ],
    timeline_variables: conditions, // this should use *my_param* after staircase procedure ?!
    sample: {
        type: 'fixed-repetitions',
        size: 10,
    }
};
```

**Reply by dokato** [reply]
I think that I cracked it down, but your code @legatrix inspired me, thanks a lot! It's just enough to put my conditions as functions too.:

```js
conditions = [
{..., diff: () => jsPsych.data.get().select('my_param').values[0] + 1},
{..., diff: () => jsPsych.data.get().select('my_param').values[0] - 2},
...
]
```

then in block:

```js
var block = {
    difficulty: jsPsych.timelineVariable('diff')()
   ....
};
```

And it's safe to use procedure mode this way!

But still it looks like a terrible hack to me. Would be could if a future version of jsPsych would introduce some remedy ;)

**Reply by legatrix** [reply]
Great! Yes, I was going to suggest the function wrappers...and also you don't need any kind of dummy trial at the beginning, you can just save `my_param` within the staircase trial. But maybe you already realised!
